pkg/cluster/cert.go

```diff
@@ -0,0 +1,226 @@
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	seededRand, cryptErr := rand.Int(rand.Reader, big.NewInt(time.Now().UnixNano()))
+	if cryptErr != nil {
+		return nil, nil, errors.WithStack(cryptErr)
+	}
+
+	now := time.Now()
+	tmpl := x509.Certificate{
+		SerialNumber: seededRand,
+		Subject: pkix.Name{
+			CommonName:   addr,
+			Organization: []string{"opendatahub-self-signed"},
+		},
+		NotBefore:             now.UTC(),
+		NotAfter:              now.Add(time.Second * 60 * 60 * 24 * 365).UTC(),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	if ip := net.ParseIP(addr); ip != nil {
+		tmpl.IPAddresses = append(tmpl.IPAddresses, ip)
+	} else {
+		if strings.HasPrefix(addr, "*.") {
+			tmpl.DNSNames = append(tmpl.DNSNames, addr[2:])
+		}
+		tmpl.DNSNames = append(tmpl.DNSNames, addr)
+	}
+
+	tmpl.DNSNames = append(tmpl.DNSNames, "localhost")
+
+	certDERBytes, err := x509.CreateCertificate(rand.Reader, &tmpl, &tmpl, key.Public(), key)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	certificate, err := x509.ParseCertificate(certDERBytes)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	certBuffer := bytes.Buffer{}
+	if err := pem.Encode(&certBuffer, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certificate.Raw,
+	}); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	keyBuffer := bytes.Buffer{}
+	if err := pem.Encode(&keyBuffer, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	}); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	return certBuffer.Bytes(), keyBuffer.Bytes(), nil
+}
+
+// GetDefaultIngressCertificate copies ingress cert secrets from openshift-ingress ns to given namespace.
+func GetDefaultIngressCertificate(ctx context.Context, c client.Client, knativeSecret, namespace string) error {
+	// Add IngressController to scheme
+	runtime.Must(operatorv1.Install(c.Scheme()))
+	defaultIngressCtrl, err := FindAvailableIngressController(ctx, c)
+	if err != nil {
+		return fmt.Errorf("failed to get ingress controller: %w", err)
+	}
+
+	defaultIngressCertName := GetDefaultIngressCertSecretName(defaultIngressCtrl)
+
+	defaultIngressSecret, err := getSecret(ctx, c, "openshift-ingress", defaultIngressCertName)
+	if err != nil {
+		return err
+	}
+
+	return copySecretToNamespace(ctx, c, defaultIngressSecret, knativeSecret, namespace)
+}
+
+func FindAvailableIngressController(ctx context.Context, c client.Client) (*operatorv1.IngressController, error) {
+	defaultIngressCtrlList := &operatorv1.IngressControllerList{}
+	listOpts := []client.ListOption{
+		client.InNamespace("openshift-ingress-operator"),
+	}
+
+	err := c.List(ctx, defaultIngressCtrlList, listOpts...)
```

Comment:
I somehow thought this was going to be changed to a `Get` by name.
Picking the first item in the list will work as long as only the default controller is there. But if users have created their own ingresses, this may no longer pick the OpenShift default one.

Reply:
If that's the expectation to just get from `default` instance, I will update this.

Reply:
Yes, IMO always getting the `default` one should be the expectation.